### PR TITLE
feat(utils): add retryOnConflict function

### PR DESCRIPTION
**Pull request by jimmythomson**
Utility function for repeatedly calling a function until a 409 Conflict error is not received

### CODE REVIEW

**Review by wboumans — APPROVED** [review]

File: src/test/utils/calculate-delay.test.ts [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+
+  describe('jitter enabled', () => {
+    // Note that these tests expect that when `calculateDelay` is
+    // called `MAX_ITERATIONS` times, it won't return a list of the
+    // exact same numbers. Theoretically, however, that scenario is
+    // possible but the chances of this are astronomically small,
+    // and so the following expectation should never fail:
+    // `expect(uniqueCount).not.toBe(results.length)`
+
+    const MAX_ITERATIONS = 500
```
Comment (jameswiltshire) [inline src/test/utils/calculate-delay.test.ts:61]: 😄 really nice way to test the jitter functionality

File: src/lib/utils/retry-on-conflict.ts [hunk 2]
```diff
@@ -0,0 +1,60 @@
+import { calculateDelay } from './calculate-delay'
+
+export interface RetryOnConflictParams<T = any> {
+  /**
+   * Any function that returns a promise
+   *
+   * This function is expected to be able to throw a {@see CommercetoolsError}
```
Comment (jameswiltshire) [inline src/lib/utils/retry-on-conflict.ts:7]: This is all probably applicable to any API we're invoking that implements proper HTTP optimistic locking with CONFLICT responses. I think commercetools is the only one for now that we need to worry about - but all of this util logic is valuable beyond just commercetools

File: src/lib/utils/retry-on-conflict.ts [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+   */
+  jitter?: boolean
+}
+
+/**
+ * Repeatedly executes a function until we get a non-409 HTTP response code
+ */
+export async function retryOnConflict<T = any>(options: RetryOnConflictParams): Promise<T> {
+  const maxRetries = options?.maxRetries || 3
+  const delayMs = options?.delayMs || 100
```
Comment (jameswiltshire) [inline src/lib/utils/retry-on-conflict.ts:40]: (really minor) could pull these out to `DEFAULT_MAX_RETRIES` constants etc

File: src/lib/utils/retry-on-conflict.ts [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+  for (let attemptCount = 1; attemptCount <= maxRetries; attemptCount++) {
+    if (attemptCount > 1) {
+      const delay = calculateDelay(attemptCount - 1, { delayMs, maxRetries, jitter })
+      await new Promise((resolve) => setTimeout(resolve, delay))
+    }
+    try {
+      return await options.executeFn(attemptCount)
+    } catch (e: any) {
+      if (!e.isCommercetoolsError || e.status !== 409 || attemptCount === maxRetries) {
```
Comment (jameswiltshire) [inline src/lib/utils/retry-on-conflict.ts:52]: (minor) could use HttpHeaders.CONFLICT

File: src/lib/utils/calculate-delay.ts [hunk 5]
```diff
@@ -0,0 +1,26 @@
+import { CommercetoolsRetryConfig } from '../api'
+
+/**
+ * Calculate how long to delay before running the request.
+ * For each retry attempt, we increase the time that we delay for.
+ */
+export function calculateDelay(retryCount: number, retryConfig?: CommercetoolsRetryConfig) {
```
Comment (jameswiltshire) [inline src/lib/utils/calculate-delay.ts:7]: (minor) `calculateRetryDelay` maybe more appropriate now this lives in its own utils package

...also same comment as earlier that the retry delay / retry config isn't really specific to commercetools

**Review by jameswiltshire — APPROVED** [review]
Fantastic stuff. Exactly the sort of reusable functionality we need to add robustness across all of our code.

File: src/lib/utils/calculate-delay.ts (hunk 5 above)
Comment (jimmythomson) [reply src/lib/utils/calculate-delay.ts:7]: Knowing what the package should be named and what it's purpose is probably the hardest part of that.

**Review by eugeingram — APPROVED** [review]

File: src/lib/utils/retry-on-conflict.ts (hunk 3 above)
Comment (jimmythomson) [reply src/lib/utils/retry-on-conflict.ts]: Done

File: src/lib/utils/retry-on-conflict.ts (hunk 4 above)
Comment (jimmythomson) [reply src/lib/utils/retry-on-conflict.ts]: Done

**Pull request merged**